From: Mike Travis <travis@sgi.com>
Date: Fri, 5 Sep 2008 14:40:21 -0700
Subject: [PATCH] x86: reduce stack requirements for send_call_func_ipi
References: bnc#425240 FATE304266
Patch-mainline: 2.6.28

* By converting the internal x86 smp_ops function send_call_func_ipi
    to pass a pointer to the cpumask_t variable, we greatly reduce the
    stack space required when NR_CPUS=4096.

    Further reduction will be realized when the send_IPI_mask interface
    is changed in 2.6.28.

Signed-off-by: Mike Travis <travis@sgi.com>
Signed-off-by: Ingo Molnar <mingo@elte.hu>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
Signed-off-by: Thomas Renninger <trenn@suse.de>
Automatically created from "patches.arch/x86_sgi_cpus4096-02-fix-send_call_func_ip.patch" by xen-port-patches.py

Index: head-2008-11-04/arch/x86/kernel/smp-xen.c
===================================================================
--- head-2008-11-04.orig/arch/x86/kernel/smp-xen.c	2008-11-04 11:55:22.000000000 +0100
+++ head-2008-11-04/arch/x86/kernel/smp-xen.c	2008-11-04 11:59:11.000000000 +0100
@@ -126,9 +126,9 @@ void xen_send_call_func_single_ipi(int c
 	send_IPI_mask(cpumask_of_cpu(cpu), CALL_FUNC_SINGLE_VECTOR);
 }
 
-void xen_send_call_func_ipi(cpumask_t mask)
+void xen_send_call_func_ipi(const cpumask_t *mask)
 {
-	send_IPI_mask(mask, CALL_FUNCTION_VECTOR);
+	send_IPI_mask(*mask, CALL_FUNCTION_VECTOR);
 }
 
 static void stop_this_cpu(void *dummy)
Index: head-2008-11-04/include/asm-x86/mach-xen/asm/smp.h
===================================================================
--- head-2008-11-04.orig/include/asm-x86/mach-xen/asm/smp.h	2008-11-04 11:55:22.000000000 +0100
+++ head-2008-11-04/include/asm-x86/mach-xen/asm/smp.h	2008-11-04 14:07:28.000000000 +0100
@@ -57,7 +57,7 @@ struct smp_ops {
 	void (*smp_send_stop)(void);
 	void (*smp_send_reschedule)(int cpu);
 
-	void (*send_call_func_ipi)(cpumask_t mask);
+	void (*send_call_func_ipi)(const cpumask_t *mask);
 	void (*send_call_func_single_ipi)(int cpu);
 };
 
@@ -106,7 +106,7 @@ static inline void arch_send_call_functi
 
 static inline void arch_send_call_function_ipi(cpumask_t mask)
 {
-	smp_ops.send_call_func_ipi(mask);
+	smp_ops.send_call_func_ipi(&mask);
 }
 
 void native_smp_prepare_boot_cpu(void);
@@ -118,13 +118,13 @@ int native_cpu_up(unsigned int cpunum);
 
 void xen_smp_send_stop(void);
 void xen_smp_send_reschedule(int cpu);
-void xen_send_call_func_ipi(cpumask_t mask);
+void xen_send_call_func_ipi(const cpumask_t *mask);
 void xen_send_call_func_single_ipi(int cpu);
 
 #define smp_send_stop		xen_smp_send_stop
 #define smp_send_reschedule	xen_smp_send_reschedule
 #define arch_send_call_function_single_ipi	xen_send_call_func_single_ipi
-#define arch_send_call_function_ipi		xen_send_call_func_ipi
+#define arch_send_call_function_ipi(m)		xen_send_call_func_ipi(&(m))
 
 #endif /* CONFIG_XEN */
 
